vmdb/app/helpers/api_helper/renderer.rb

```diff
@@ -291,8 +291,8 @@ def attr_accessible?(object, attr)
     def attr_virtual?(object, attr)
       primary = attr_split(attr).first
       return false unless object && object.respond_to?(:attributes) && object.respond_to?(primary)
-      is_reflection = object.class.reflections_with_virtual.keys.collect(&:to_s).include?(primary)
-      is_virtual_column = object.class.virtual_columns_hash.keys.include?(primary) unless is_reflection
+      is_reflection     = cached_class_reflections(object).include?(primary)
+      is_virtual_column = cached_class_virtual_columns(object).include?(primary) unless is_reflection
```

Comment:
Let's not introduce another layer of caching because the virtual methods there are already cached.  The reason the methods above are slow is that they do `.keys.include?` turning an O(1) into an O(n).  Then this method is called in a loop for every attribute passed, making it exponentially slow.  You can just do `.include?` or `.key?` directly on the hash to get back to O(1).

There is no `.to_sym` security issue because you are guarded by the respond_to? check.